components/tidb_query/src/codec/convert.rs

```diff
@@ -710,8 +710,8 @@ pub fn pad_zero_for_binary_type(s: &mut Vec<u8>, ft: &FieldType) {
         return;
     }
     let flen = flen as usize;
-    let tp = ft.as_accessor().tp();
-    if tp == FieldTypeTp::String && ft.is_binary_string_like() && s.len() < flen {
+    if ft.as_accessor().tp() == FieldTypeTp::String && ft.is_binary_string_like() && s.len() < flen
```

Comment:
Maybe you want `ft.collation() == Collation::Binary`?

is_binary_string_like stands for `is_binary` && `is_string_like`. But you only want `is_binary` && `is_string`.

Reply:
Yes, because we judge `ft.as_accessor().tp() == FieldTypeTp::String` first, then the `IsString(ft.Tp)` will always be true.
```go
func IsBinaryStr(ft *FieldType) bool {
	return ft.Collate == charset.CollationBin && IsString(ft.Tp)
}
```

I will fix this.